File: Data Structures & Algorithms/rotting-fruit/submission-5.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        r=len(grid)
        c=len(grid[0])
        queue=deque()
        seen=set()
        directions=[(0,1),(0,-1),(1,0),(-1,0)]
        time=0
        fresh=0
        def valid(row,col):
            nonlocal r
            nonlocal c
            nonlocal grid
            return 0<=row<r and 0<=col<c and grid[row][col]==1
        for i in range(r):
            for j in range(c):
                if grid[i][j]==2:
                    queue.append((i,j))
                    seen.add((i,j))
                if grid[i][j]==1:
                    fresh+=1
        if len(queue)==0:
            if fresh==0:
                return 0
            else:
                return -1
        while len(queue)>0:
            if fresh==0:
                break
            for i in range(len(queue)):
                curr=queue.popleft()
                for d in directions:
                    dx,dy=d
                    row,col=curr
                    new_row=row+dx
                    new_col=col+dy
                    if valid(new_row,new_col):
                        queue.append((new_row,new_col))
                        grid[new_row][new_col]=2
                        fresh-=1
            time+=1
        if fresh==0:
            return time
        else:
            return -1
        
        

File: Data Structures & Algorithms/path-with-minimum-effort/submission-5.java
class Solution {
    public boolean valid(int[][] grid, int row, int col){
        int r=grid.length;
        int c=grid[0].length;
        return((0<=row)&&(row<r)&&(0<=col)&&(col<c));
    }
    public int minimumEffortPath(int[][] heights) {
        int r=heights.length;
        int c=heights[0].length;
        int[][] directions=new int[][]{{0,1},{1,0},{-1,0},{0,-1}};
        PriorityQueue<int[]> pq = new PriorityQueue<>((a,b)->a[0]-b[0]);
        int[] head=new int[]{0,0,0};
        pq.offer(head);
        int[][] shortest=new int[r][c];
        for(int i=0;i<r;i++){
            for (int j=0;j<c;j++){
                shortest[i][j]=Integer.MAX_VALUE;
            }
        }
        while (!pq.isEmpty()){
            int[] curr=pq.poll();
            int w1=curr[0];
            int r1=curr[1];
            int c1=curr[2];
            if (shortest[r1][c1]!=Integer.MAX_VALUE){
                continue;
            }
            shortest[r1][c1]=w1;
            for(int[] dir: directions){
                int newRow=r1+dir[0];
                int newCol=c1+dir[1];
                if (valid(heights,newRow,newCol)){
                    int[] temp=new int[]{Math.max(w1,Math.abs(heights[newRow][newCol]-heights[r1][c1])),newRow,newCol};
                    pq.offer(temp);
                }
            }
        }
        return shortest[r-1][c-1];        
    }
}